content/posts/2023-11-20-tools-for-learning-llvm-tablegen.md

```diff
@@ -0,0 +1,436 @@
+hypothetical backend could) so there is no option to compile to binary or
+execute code.
+
+By default you will see the records printed as plain text. A backend can be
+chosen by adding a compiler option manually or as an “Action” from the
+“Overrides” menu.
+
+It’s important to note that TableGen backends have very specific expectations of
+what will be in the source code. Almost as if you had a C++ compiler that
+wouldn’t compile for Arm unless it saw the symbol `arm_is_cool` somewhere in the
+source code.
+
+In the LLVM monorepo all the required classes are set up for you, but in
+Compiler Explorer they are not. So if you want to experiment with an existing
+backend I suggest providing stub implementations of the classes, or copying some
+in from the LLVM project. Standard includes from `include/llvm/*.td` can also be
+used.
+
+Developing a backend within Compiler Explorer is not possible at this time, but
+you can select the JSON backend and copy that JSON to give to local scripts.
+
+Multi-file projects (“IDE mode”) also work as expected, so you can have your own
+[include files](https://godbolt.org/z/4qhdoaMjE) if you like.
+
+Finally, remember that Compiler Explorer examples can be shared. If you’re
+asking or answering questions about TableGen, always include a Compiler Explorer
+link if you can!
+
+# Jupyter Notebooks
+
+[Jupyter](https://jupyter.org/) creates interactive notebooks that combine text
+and code into one document. With the ability to edit that code rerun it to
+update the results in the notebook.
+
+This is great for taking notes or building up large examples from small chunks
+of code. Then you can export your notes as a notebook anyone can edit, or in
+static formats like PDF or Markdown.
+
+TableGen can be used in one of these notebooks via the
+[TableGen Jupyter Kernel](https://github.com/llvm/llvm-project/tree/main/llvm/utils/TableGen/jupyter).
+Install instructions are on the linked page and you can watch me talk more about
+it [here](https://www.youtube.com/watch?v=Gf0FUiY2TRo).
+
+**Note:** There’s also an [MLIR kernel](https://github.com/llvm/llvm-project/tree/main/mlir/utils/jupyter)
+for Jupyter, amongst many, many others.
+
+We’ve aimed to give the same experience as other languages, so I will focus not
+on how to use a notebook, but instead on what we’ve been able to make with them.
+
+## TableGen Tutorial
+
+The first notebook is an
+[introduction to TableGen](https://github.com/llvm/llvm-project/blob/main/llvm/utils/TableGen/jupyter/tablegen_tutorial_part_1.ipynb).
+You can read it on GitHub or download for use in Jupyter.
+
+The great thing about this is that if something in that document doesn’t make
+sense to you, you can edit the examples and experiment. Maybe it glosses over a
+part you find really interesting. You can add a new code and see what happens.
+
+## How to Write a TableGen Backend
+
+This notebook is different. The language is Python and instead of focusing on
```

Comment:
"This notebook" doesn't flow too well from "How to Write a TableGen Backend". It would help if it was somewhat clearer that we are iterating over in-tree examples here. Perhaps "Example 1" and "Example 2"? Or something similar.

Reply:
I really want the titles to describe what's in each one, so I've added "Notebook" onto the end of them.